From 0fb21777a4b954771e71b6c8255c384a2abf9dbe Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Fri, 11 Feb 2022 07:30:58 +0200
Subject: [PATCH] [SW-72556] greco: change legacy boot detection logic

Due to DCCM ECC issue in arcboot rom, hybrid mode is enabled
for greco where execution starts from BTL and BTL starts
ROM flow on ARC1. Due to this, BTL_EN bit is set but still
ROM flow is executed. In mixed flow, BTL version needs to
be passed to host. This patch uses security_bypass strap
bit to distinguish between mixed flow and pure arcboot flow.

Change-Id: I7243666cb1b4494be45f673dff0d7d096f6ca92a
---

diff --git a/arch/greco/arc/common.c b/arch/greco/arc/common.c
index 4788c01..b77ceb7 100644
--- a/arch/greco/arc/common.c
+++ b/arch/greco/arc/common.c
@@ -165,12 +165,14 @@
 	 * Due to DCCM ECC issue in arcboot rom, hybrid mode is enabled
 	 * for greco where execution starts from BTL and BTL starts
 	 * ROM flow on ARC1. Due to this, BTL_EN bit is set but still
-	 * ROM flow is executed. Since its mixed flow, we
-	 * will pass btl version to host
+	 * ROM flow is executed. In mixed flow, BTL version needs to
+	 * be passed to host. To distinguish between mixed flow and
+	 * arcboot flow Security bypass strap is used.
 	 */
-	u32 boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L);
+	u32 boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS_H);
 
-	return boot_strap & PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L_BTL_EN_MASK;
+	return !(boot_strap &
+		 PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_H_SECURITY_BYPASS_MASK);
 }
 
 #define SP_SRAM_60K_SZ_16B_GRANULARITY	0xF00
